File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  ans = l = 0
  r = 1
  while r < prices.length
    ans = [ans, prices[r]-prices[l]].max
    l = r if prices[r] < prices[l]
    r += 1
  end
  ans
end

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
  ans, l, r := 0, 0, 0
  for r < len(prices) {
    ans = max(ans , prices[r] - prices[l])
    if prices[r] < prices[l] {
      l = r
    }
    r++
  }
  return ans
}